scipy/sparse/sputils.py

```diff
@@ -251,6 +251,61 @@ def validateaxis(axis):
             raise ValueError("axis out of range")
 
 
+def check_shape(args, current_shape=None):
+    """Imitate numpy.matrix handling of shape arguments"""
+    int_types = (int, np.int8, np.int16, np.int32, np.int64,
+                 np.uint8, np.uint16, np.uint32, np.uint64,
+                 np.intc, np.intp)
+
+    def check_int(arg):
+        if isinstance(arg, int_types):
+            return int(arg)
+        elif isinstance(arg, np.ndarray) and np.isscalar(arg) and \
+                arg.dtype in int_types:
+            return int(arg)
+        else:
+            raise TypeError("'{}' object cannot be interpreted as an "
+                            "integer".format(args[0].__class__.__name__))
+
+    if len(args) == 0:
+        raise TypeError("'reshape' missing 1 required positional argument: "
+                        "'shape'")
+    elif len(args) == 1:
+        try:
+            shape_iter = iter(args[0])
+        except TypeError:
+            new_shape = (check_int(args[0]), )
+        else:
+            new_shape = tuple(check_int(arg) for arg in shape_iter)
+    else:
+        new_shape = tuple(check_int(arg) for arg in args)
+
+    # Check the current size only if needed
+    if current_shape is not None:
+        current_size = np.prod(current_shape, dtype=int)
+        new_size = np.prod(new_shape, dtype=int)
+        if new_size != current_size:
+            raise ValueError('cannot reshape array of size {} into shape {}'
+                             .format(new_size, new_shape))
+
+    # Add and remove ones like numpy.matrix.reshape
+    if len(new_shape) != 2:
+        new_shape = tuple(arg for arg in new_shape if arg != 1)
+
+        if len(new_shape) == 0:
+            new_shape = (1, 1)
+        elif len(new_shape) == 1:
+            new_shape = (1, new_shape[0])
+        elif len(new_shape) > 2:
+            raise ValueError('shape too large to be a matrix')
+
+    # Check for negatives
+    if new_shape[0] < 0 or new_shape[1] < 0:
+        raise ValueError("'shape' elements cannot be negative")
```

Comment:
We should support a single `-1` dimension: `foo.reshape((5, -1))`

Reply:
I forgot about this case. I'll have to change some of the sparse tests, because they expect an error.